sql/src/main/java/io/crate/planner/operators/InsertFromValues.java

```diff
@@ -0,0 +1,766 @@
+import io.crate.execution.engine.indexing.ShardLocation;
+import io.crate.execution.engine.indexing.ShardedRequests;
+import io.crate.execution.support.RetryListener;
+import io.crate.expression.InputFactory;
+import io.crate.expression.InputRow;
+import io.crate.expression.symbol.Assignments;
+import io.crate.expression.symbol.SelectSymbol;
+import io.crate.expression.symbol.Symbol;
+import io.crate.metadata.IndexParts;
+import io.crate.metadata.Reference;
+import io.crate.metadata.RelationName;
+import io.crate.metadata.doc.DocTableInfo;
+import io.crate.metadata.table.Operation;
+import io.crate.metadata.tablefunctions.TableFunctionImplementation;
+import io.crate.planner.DependencyCarrier;
+import io.crate.planner.ExecutionPlan;
+import io.crate.planner.PlannerContext;
+import io.crate.types.DataType;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.admin.indices.create.CreatePartitionsRequest;
+import org.elasticsearch.action.admin.indices.create.TransportCreatePartitionsAction;
+import org.elasticsearch.action.bulk.BackoffPolicy;
+import org.elasticsearch.action.support.master.AcknowledgedResponse;
+import org.elasticsearch.cluster.metadata.MetaData;
+import org.elasticsearch.cluster.routing.ShardIterator;
+import org.elasticsearch.cluster.routing.ShardRouting;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.io.stream.NotSerializableExceptionWrapper;
+import org.elasticsearch.index.IndexNotFoundException;
+
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.stream.StreamSupport;
+
+import static io.crate.data.SentinelRow.SENTINEL;
+import static io.crate.execution.engine.indexing.ShardingUpsertExecutor.BULK_REQUEST_TIMEOUT_SETTING;
+
+
+public class InsertFromValues implements LogicalPlan {
+
+    private static final BackoffPolicy BACK_OFF_POLICY = LimitedExponentialBackoff.limitedExponential(1000);
+
+    private final TableFunctionRelation tableFunctionRelation;
+    private final ColumnIndexWriterProjection writerProjection;
+
+    InsertFromValues(TableFunctionRelation tableFunctionRelation,
+                     ColumnIndexWriterProjection writerProjection) {
+        this.tableFunctionRelation = tableFunctionRelation;
+        this.writerProjection = writerProjection;
+    }
+
+    @Override
+    public StatementType type() {
+        return StatementType.INSERT;
+    }
+
+    @Override
+    public void execute(DependencyCarrier dependencies,
```

Comment:
`execute` and `executeBulk` share some input creation logic, etc. not sure how it can be deduplicated. Also, it is a bit hard to have some generic logic for creating and validation logic for those methods due the way we handle/bind on duplicated assignment symbols.